pydantic_ai_slim/pydantic_ai/models/anthropic.py

```diff
@@ -663,26 +652,91 @@ def _add_builtin_tools(
                 )
         return tools, mcp_servers, beta_features
 
-    def _infer_tool_choice(
+    def _infer_tool_choice(  # noqa: C901
         self,
-        tools: list[BetaToolUnionParam],
         model_settings: AnthropicModelSettings,
         model_request_parameters: ModelRequestParameters,
-    ) -> BetaToolChoiceParam | None:
-        if not tools:
-            return None
+    ) -> tuple[list[BetaToolUnionParam], BetaToolChoiceParam | None]:
+        """Determine which tools to send and the API tool_choice value.
+
+        Returns:
+            A tuple of (filtered_tools, tool_choice).
+        """
+        thinking_enabled = model_settings.get('anthropic_thinking') is not None
+        function_tools = model_request_parameters.function_tools
+        output_tools = model_request_parameters.output_tools
+
+        resolved = _resolve_tool_choice(model_settings, model_request_parameters)
+
+        if resolved is None:
+            tool_defs_to_send = [*function_tools, *output_tools]
```

Comment:
This is the same thing that `resolved.filter_tools` returns for `'auto'`, so we can likely dedupe this by treating `None` and `'auto'` the same at another level